### ISSUE TITLE
non-mail URL containing valid email (sometimes) opens compose dialog

### ISSUE THREAD

**Issue by Bogidon** [nylas/N1#2071]
This issue happens when sending links that contain email addresses _as plaintext_. e.g. `https://www.google.com/search?q=test@example.com`. When clicking the link, N1 will open a compose dialog instead of letting the system handle the link.

This is impossible to do with N1 (as far as I know) because N1 will convert URLs to proper links. e.g. `[https://www.google.com/search?q=test@example.com](https://www.google.com/search?q=test%40example.com)`.

To test this, send https://www.google.com/search?q=test@example.com to yourself from a client that allows plaintext or html editing.

This may have to do with #2061 or #403

##### What operating system are you using?
OS X 10.11.4

##### What version of N1 are you using?
0.4.25-a22631a

**Comment by Bogidon** [comment]
I guess this is why they say to not use Regex when parsing emails. Not sure how you could solve it though. Maybe check for `://`?

**Comment by Bogidon** [comment]
It looks like checking if there are any of [these characters](https://tools.ietf.org/html/rfc5322#section-3.2.3): `(),:;<>@[\]` present without being surrounded by quotes could invalidate the email.

**Comment by bengotow** [comment]
Thanks for filing this—I've been able to reproduce it with your example. You know, the more I work on this project the more I realize a mail client is really just a huge pile of regex expressions. :-)

I think the problem is that we're running email detection before link detection, and the email regexp does not require that the email is a whole word (surrounded by spaces, etc.). I think making the email regexp more restrictive will only partially solve the problem. It'd still the `test@example.com` part of the URL into an email address, causing the URL (evaluated later) to then just be ` https://www.google.com/search?q=`. 

I think the best solution may be to run the URL linker first, but right now it considers `bengotow@gmail.com` to be a valid URL (technically I think it's the abbreviated version of `http://bengotow:@gmail.com`)

**Comment by bengotow** [comment]
Here's my proposed solution:

1. I've added a manual exclusion test to the email matcher, which prevents the following specific pattern from matching as an email address: `/\..*[\/|\?].*\@/`. This means that `gmail.com?q=bengotow@gmail.com` will not match, and neither will `gmail.com/bengotow@gmail.com`. These characters /are/ valid for emails, but they are also valid for domains, and when we find all of them in that particular order, we'll just assume you meant a domain.

2. I've added a bit of logic to the autolinker so that if two expressions match the same node, it will apply the one that matches more text. So in the case of `https://www.google.com/search?q=test@example.com`, the URL matcher matches a larger block of text and will be preferred.

I also added a new spec for this.
